### PR TITLE
[SPARK-42702][SPARK-42623][SQL] Support parameterized query in subquery and CTE

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This PR fixes a few issues of parameterized query:
1. replace placeholders in CTE/subqueries
2. don't replace placeholders in non-DML commands as it may store the original SQL text with placeholders and we can't resolve it later (e.g. CREATE VIEW).

### Why are the changes needed?

make the parameterized query feature complete

### Does this PR introduce _any_ user-facing change?

yes, bug fix

### How was this patch tested?

new tests

### CODE REVIEW

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

File: core/src/main/resources/error/error-classes.json
Comment (srielau) [inline core/src/main/resources/error/error-classes.json:1735]: Is "query parameters" a term? I thought we go by "parameter marker". Either way we can be more descriptive.

```suggestion
          "Parameter markers in a DDL statement. Parameter markers must only be used in a query, or DML statement. Usage within CREATE or ALTER is unsupported."
```

File: core/src/main/resources/error/error-classes.json
Comment (srielau) [inline core/src/main/resources/error/error-classes.json:1733]:
```suggestion
      "PARAMETERIZED_DDL" : {
```

**Review by srielau — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:50]: If you don't mind, can we have a negative test case for this **case-sensitive**?

**Review by dongjoon-hyun — COMMENTED** [review]
Also, cc @sunchao , @huaxingao , @viirya , too.

**Comment by cloud-fan** [comment]
The failed `BasicSchedulerIntegrationSuite` is not related to this PR, I'm merging it to master/3.4, thanks for the review!

**Comment by cloud-fan** [comment]
This is a bug fix of a new feature in 3.4, so I won't call it a release blocker. I've set the fixed version to 3.4.0, if rc3 passes, I'll change it to 3.4.1.

**Comment by HyukjinKwon** [comment]
Seems like the compliation didn't pass. Let me just quickly revert this and reopen.

**Comment by cloud-fan** [comment]
maybe there is a conflict right after my last commit, let me rebase

**Review by viirya — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending CIs)

**Comment by cloud-fan** [comment]
GA passes, let me merge it back.

**Pull request closed without merging**
